### PR TITLE
Fix error's stacktrace erasing

### PR DESCRIPTION
**Pull request by sergioboi**

### CODE REVIEW

File: subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/InHouseInstrumentationTestRunner.kt [hunk 1]
```diff
@@ -210,7 +209,7 @@ abstract class InHouseInstrumentationTestRunner :
 
         testRunEnvironment.executeIfRealRun {
             Espresso.setFailureHandler(
-                ReportFriendlyFailureHandler(DefaultFailureHandler(targetContext))
```
Comment (sergioboi) [inline subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/InHouseInstrumentationTestRunner.kt:213]: DefaultFailure handler does not do any improvements to stack traces.
But erase stack traces
`error.setStackTrace(Thread.currentThread().getStackTrace());`
Look at `androidx.test.espresso.base.DefaultFailureHandler#getUserFriendlyError` for details

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt [hunk 2]
```diff
@@ -50,50 +49,48 @@ class ReportFriendlyFailureHandler(private val defaultHandler: FailureHandler) :
...
-                        .withCause(e)
+        throw when {
+            error.isCausedBy<AppNotIdleException> { it.message.orEmpty().contains("ASYNC_TASKS_HAVE_IDLED") } -> {
+                val asyncTaskThreads = Thread.getAllStackTraces()
+                    .filterKeys { it.name.contains("AsyncTask") }
+                val threadDump = asyncTaskThreads.entries.joinToString(separator = "\n") { it.dumpState() }
+                val exception = createExceptionWithPrivateStringConstructor<AppNotIdleException>(
+                    "AsyncTask is still running in background. Thread dump:\n$threadDump"
+                )
+                exception.initCause(error)
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt:60]: Before we creating Exception but lost the original cause

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt [hunk 3]
```diff
@@ -50,50 +49,48 @@ class ReportFriendlyFailureHandler(private val defaultHandler: FailureHandler) :
...
+                exception
+            }
+            error is PerformException -> {
+                // RecyclerView descendant checks implemented via ViewActions (gross)
+                if (error.actionDescription.startsWith("Check descendant view")) {
+                    AssertionFailedError("Не прошла проверка: ${error.actionDescription}").apply { initCause(error.cause) }
+                } else {
+                    PerformException.Builder()
+                        .from(error)
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt:79]: Before we adding cause but `from` do everything for us

File: subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt [hunk 4]
```diff
@@ -50,50 +49,48 @@ class ReportFriendlyFailureHandler(private val defaultHandler: FailureHandler) :
...
+            error is AssertionError -> AssertionFailedError(error.normalizedMessage()).apply { initCause(error.cause) }
+            error is NoMatchingRootException -> error.toNormalizedException()
+            error is NoMatchingViewException -> error.toNormalizedException()
+            else -> error
         }
     }
 
     private inline fun <reified T : Exception> T.toNormalizedException(): T {
-        return createExceptionWithPrivateStringConstructor(this.normalizedMessage())
+        val exception = createExceptionWithPrivateStringConstructor<T>(this.normalizedMessage())
```
Comment (sergioboi) [inline subprojects/android-test/test-report/src/main/kotlin/com/avito/android/test/report/ReportFriendlyFailureHandler.kt:92]: Before we normalized the error message but lost an original stack trace

File: subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/InHouseInstrumentationTestRunner.kt (hunk 1 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/test-inhouse-runner/src/main/kotlin/com/avito/android/runner/InHouseInstrumentationTestRunner.kt:213]: It'd good to file an issue about `DefaultFailureHandler`

**Review by eugene-krivobokov — APPROVED** [review]

**Pull request merged**